pkg/resources/management.cattle.io/v3/project/validator.go

```diff
@@ -143,6 +151,48 @@ func (a *admitter) admitCommonCreateUpdate(oldProject, newProject *v3.Project) (
 	return admission.ResponseAllowed(), nil
 }
 
+func (a *admitter) validateContainerDefaultResourceLimit(limit *v3.ContainerResourceLimit) error {
+	if limit == nil {
+		return nil
+	}
+	fieldPath := projectSpecFieldPath.Child(containerLimitField)
+	requestsCPU, err := parseResource(limit.RequestsCPU)
+	if err != nil {
+		return field.Invalid(fieldPath, limit.RequestsCPU, fmt.Sprintf("failed to parse container default requested CPU: %s", err))
+	}
+	limitCPU, err := parseResource(limit.LimitsCPU)
+	if err != nil {
+		return field.Invalid(fieldPath, limit.LimitsCPU, fmt.Sprintf("failed to parse container default CPU limit: %s", err))
+	}
+	requestsMemory, err := parseResource(limit.RequestsMemory)
+	if err != nil {
+		return field.Invalid(fieldPath, limit.RequestsMemory, fmt.Sprintf("failed to parse container default requested memory: %s", err))
+	}
+	limitMemory, err := parseResource(limit.LimitsMemory)
+	if err != nil {
+		return field.Invalid(fieldPath, limit.LimitsMemory, fmt.Sprintf("failed to parse container default memory limit: %s", err))
+	}
+	resourceList := v1.ResourceList{
+		v1.ResourceRequestsCPU:    requestsCPU,
+		v1.ResourceLimitsCPU:      limitCPU,
+		v1.ResourceRequestsMemory: requestsMemory,
+		v1.ResourceLimitsMemory:   limitMemory,
+	}
+	if negativeResources := quota.IsNegative(resourceList); len(negativeResources) > 0 {
+		return field.Invalid(fieldPath, limit,
+			fmt.Sprintf("resource requests and limits can't have negative values, adjust the request/limit configurations for the following resources: %v", negativeResources))
+	}
+	if !requestsCPU.IsZero() && !limitCPU.IsZero() && requestsCPU.Cmp(limitCPU) > 0 {
+		fieldErr := field.Invalid(fieldPath, limit, fmt.Sprintf("requested CPU %s is greater than limit %s", limit.RequestsCPU, limit.LimitsCPU))
+		err = errors.Join(err, fieldErr)
+	}
+	if !requestsCPU.IsZero() && !limitCPU.IsZero() && requestsMemory.Cmp(limitMemory) > 0 {
```

Comment:
Good catch! The reason for all this is the difference between a literal "0" in limits or requests and absence of a value. Kubernetes treats a "0" as something restrictive. So I decided against setting a limit or request to "0" if the value in the request is an empty string. Using "0" here wouldn't matter because this is webhook, it only inspects a request, and nothing would literally set anything to "0", but I didn't want to operate with these misleading terms here.

Yet I am using this `IsZero` method, which could still be slightly misleading...

But the main check is needed here because it's OK to provide only request or only a limit. So we don't want to reject such configurations.

Reply:
The tests didn't catch this because I mostly had both CPU and memory specified, I adjusted the tests to make them a bit more varied.